src/GafferScene/BranchCreator.cpp

```diff
@@ -82,30 +93,61 @@ const Gaffer::StringPlug *BranchCreator::parentPlug() const
 	return getChild<StringPlug>( g_firstPlugIndex );
 }
 
-Gaffer::ObjectPlug *BranchCreator::mappingPlug()
+Gaffer::PathMatcherDataPlug *BranchCreator::filteredPathsPlug()
 {
-	return getChild<ObjectPlug>( g_firstPlugIndex + 1 );
+	return getChild<PathMatcherDataPlug>( g_firstPlugIndex + 1 );
 }
 
-const Gaffer::ObjectPlug *BranchCreator::mappingPlug() const
+const Gaffer::PathMatcherDataPlug *BranchCreator::filteredPathsPlug() const
 {
-	return getChild<ObjectPlug>( g_firstPlugIndex + 1 );
+	return getChild<PathMatcherDataPlug>( g_firstPlugIndex + 1 );
+}
+
+Gaffer::PathMatcherDataPlug *BranchCreator::parentPathsPlug()
+{
+	return getChild<PathMatcherDataPlug>( g_firstPlugIndex + 2 );
+}
+
+const Gaffer::PathMatcherDataPlug *BranchCreator::parentPathsPlug() const
+{
+	return getChild<PathMatcherDataPlug>( g_firstPlugIndex + 2 );
+}
+
+Gaffer::AtomicCompoundDataPlug *BranchCreator::mappingPlug()
+{
+	return getChild<AtomicCompoundDataPlug>( g_firstPlugIndex + 3 );
+}
+
+const Gaffer::AtomicCompoundDataPlug *BranchCreator::mappingPlug() const
+{
+	return getChild<AtomicCompoundDataPlug>( g_firstPlugIndex + 3 );
 }
 
 void BranchCreator::affects( const Plug *input, AffectedPlugsContainer &outputs ) const
 {
-	SceneProcessor::affects( input, outputs );
+	FilteredSceneProcessor::affects( input, outputs );
 
 	if( input->parent<ScenePlug>() == inPlug() )
 	{
 		outputs.push_back( outPlug()->getChild<ValuePlug>( input->getName() ) );
 	}
-	else if( input == parentPlug() )
+
+	if( input == parentPlug() || input == filteredPathsPlug() )
+	{
+		outputs.push_back( parentPathsPlug() );
+	}
+
+	if( input == parentPathsPlug() )
 	{
-		outputs.push_back( mappingPlug() );
 		outputs.push_back( outPlug()->setNamesPlug() );
 	}
-	else if( input == mappingPlug() )
+
+	if( input == inPlug()->childNamesPlug() )
+	{
+		outputs.push_back( mappingPlug() );
+	}
```

Comment:
Good catch. The mapping is all about resolving name collisions, so it depends on the input child names (covered here) and the branch child names (not covered, and never was) This looks like a bug we've had all along. I don't want the derived classes to have to think about `mappingPlug()`, but what I think might make sense is a `virtual bool affectsBranchChildNames( const Plug *input ) const` method they could overload. I've been wanting to overhaul SceneNode so that the `hashObject()` and `computeObject()` etc methods have a corresponding `affectsObject()` method, and this would fit that pattern. Worth doing now you think?

Reply:
Guess it's up to you, but leaving around hashing bugs always seems to create the potential for horrible things to happen.

Reply:
This one is a dirty propagation bug rather than a hashing one, so a bit less heinous. I'll fix it now though. I have an optimisation branch where I use dirty propagation to selectively invalidate the hash cache rather than clear the whole thing, and that requires the dirty propagation to be absolutely watertight.

Reply:
Fixed in c71d976